src/database-postgresql.cpp

```diff
@@ -0,0 +1,375 @@
+#include "log.h"
+#include "filesys.h"
+
+Database_PostgreSQL::Database_PostgreSQL(Settings &conf)
+{
+	m_conf = conf;
+
+	if (PQisthreadsafe() < 1)
+		infostream << "Postgres backend is not thread safe." << std::endl;
+
+	connectString();
+	dbconnect();
+	prepare();
+}
+
+void Database_PostgreSQL::verifyDatabase()
+{
+	if (PQstatus(m_conn) == CONNECTION_BAD) {
+		PQreset(m_conn);
+			
+		if (ping())
+			return;
+		else
+			throw BaseException(PQerrorMessage(m_conn));
+	}
+
+	return;
+}
+
+void Database_PostgreSQL::prepare()
+{
+	Oid writeFormats[]  = {20, 17};
+	Oid updateFormats[] = {17, 20};
+	Oid readFormats[]   = {20};
+
+	m_res = PQprepare(m_conn, "write_block",
+			"INSERT INTO blocks (pos, data) VALUES ($1::bigint, $2::bytea)", 2,
+			(const Oid *)writeFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "update_block",
+			"UPDATE blocks SET data = $1::bytea WHERE pos = $2::bigint", 2,
+			(const Oid *)updateFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "read_block",
+			"SELECT data FROM blocks WHERE pos = $1::bigint", 1,
+			(const Oid *)readFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "delete_block",
+			"DELETE FROM blocks WHERE pos = $1::bigint", 1,
+			(const Oid *)readFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "read_pos", "SELECT pos FROM blocks", 0, NULL);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	PQclear(m_res);
+
+	infostream << "Postgres: SQL statements prepared." << std::endl;
+}
+
+void Database_PostgreSQL::dbconnect()
+{
+	m_conn = PQconnectdb(m_conninfo.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) 
+	       	throw BaseException(PQerrorMessage(m_conn));
+
+	m_serverVersion = PQserverVersion(m_conn);
+
+	if (m_serverVersion < 90300)
+	       	throw BaseException("Postgres: Server version 9.3 or greater only.");
+
+	createDatabase();
+
+	infostream << "Postgres: Server version " << m_serverVersion << std::endl;
+	infostream << "Postgres: Database connection made." << std::endl;
+}
+
+void Database_PostgreSQL::connectString()
+{
+	if (m_conninfo.length() > 1)
+		return;
+
+	std::ostringstream confstream;
+
+	// username
+	if (m_conf.exists("pg_username")) {
+	    m_user = m_conf.get("pg_username");
+	} else {
+	    m_user = "minetest";
+	    infostream << "Postgres: Using default user, minetest.\n";
+	}
+
+	// Postgres host
+	if (m_conf.exists("pg_host"))
+	    m_host = m_conf.get("host");
+
+	// Port number
+	if (m_conf.exists("pg_port") && m_conf.exists("pg_host")) {
+	    m_port = m_conf.get("pg_port");
```

Comment:
This is indented partly with tabs, partly with spaces, which is bad, use tabs only.